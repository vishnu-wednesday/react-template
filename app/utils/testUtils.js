import React from 'react';
import { render } from '@testing-library/react';
import { Provider } from 'react-redux';
import { browserHistory, BrowserRouter, MemoryRouter, Route } from 'react-router-dom';
import { ThemeProvider } from 'styled-components';
import { createIntl, createIntlCache, IntlProvider } from 'react-intl';

import configureStore from '@app/configureStore';
import { DEFAULT_LOCALE, translationMessages } from '@app/i18n';
import ConnectedLanguageProvider from '@containers/LanguageProvider';
import { IntlGlobalProvider } from '@components/IntlGlobalProvider';
import { createMemoryHistory } from 'history';

export const renderWithIntl = (children) =>
  render(
    <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
      <BrowserRouter>
        <IntlGlobalProvider>{children}</IntlGlobalProvider>
      </BrowserRouter>
    </IntlProvider>
  );

export const getComponentStyles = (Component, props = {}) => {
  renderWithIntl(Component(props));
  const { styledComponentId } = Component(props).type;
  const componentRoots = document.getElementsByClassName(styledComponentId);
  // eslint-disable-next-line no-underscore-dangle
  return window.getComputedStyle(componentRoots[0])._values;
};

export const renderProvider = (children) => {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <ConnectedLanguageProvider messages={translationMessages}>
        <ThemeProvider
          theme={{
            main: 'violet'
          }}
        >
          {children}
        </ThemeProvider>
      </ConnectedLanguageProvider>
    </Provider>
  );
};

export function renderWithRouterMatch(
  ui,
  { path = '/', route = '/', history = createMemoryHistory({ initialEntries: [route] }) } = {}
) {
  const store = configureStore({}, browserHistory).store;
  return render(
    <Provider store={store}>
      <IntlProvider locale={DEFAULT_LOCALE} messages={translationMessages[DEFAULT_LOCALE]}>
        <MemoryRouter initialEntries={[route]}>
          <Route path={path}>{ui}</Route>
        </MemoryRouter>
      </IntlProvider>
    </Provider>
  );
}

export const timeout = (ms) => new Promise((resolve) => setTimeout(resolve, ms));
export const apiResponseGenerator = (ok, data) => ({
  ok,
  data
});

export const createIntlUtil = (locale = DEFAULT_LOCALE) => {
  const cache = createIntlCache();
  const getMessages = (locale) => translationMessages[locale];

  return createIntl(
    {
      locale: locale,
      messages: getMessages(locale)
    },
    cache
  );
};
