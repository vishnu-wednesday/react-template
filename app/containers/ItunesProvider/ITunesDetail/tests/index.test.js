/* eslint-disable no-console */
/**
 *
 * Tests for HomeContainer (ITunesGrid)
 *
 */

import React from 'react';
import { rest } from 'msw';
import { setupServer } from 'msw/node';

import { timeout, renderWithRouterMatch } from '@utils/testUtils';
import { ITunesDetailTest as ITunesDetailContainer } from '../index';
import ITunesDetailContainerLazily from '../Loadable';
import { mockData } from '@app/components/TrackCard/tests/mockTrackResponse';

describe('<ITunesDetails Container Tests', () => {
  let submitSpy;
  let clearSpy;

  const server = setupServer(
    rest.get('/lookup', (req, res, ctx) => {
      const query = req.url.searchParams;
      const id = query.get('id');
      return res(
        ctx.json({
          ok: true,
          results: [
            {
              songDetails: id
            }
          ]
        })
      );
    })
  );

  beforeAll(() => server.listen());

  beforeEach(() => {
    submitSpy = jest.fn();
    clearSpy = jest.fn();
  });

  /* 
   This does test the loadable file. Very weird though.
  */

  it('should render with loadable', async () => {
    const { baseElement } = await renderWithRouterMatch(
      <ITunesDetailContainerLazily
        dispatchGetTrackDetails={submitSpy}
        dispatchClearTrackDetails={clearSpy}
        trackDetails={mockData.results[0]}
      />,

      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    expect(baseElement).toBeDefined();
  });

  it('should render and match the snapshot', () => {
    const { baseElement } = renderWithRouterMatch(
      <ITunesDetailContainer dispatchGetTrackDetails={submitSpy} dispatchClearTrackDetails={clearSpy} />,
      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    expect(baseElement).toMatchSnapshot();
  });

  it('should dispatch dispatchGetTrackDetails after rendering', async () => {
    renderWithRouterMatch(
      <ITunesDetailContainer dispatchGetTrackDetails={submitSpy} dispatchClearTrackDetails={clearSpy} />,
      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    await timeout(200);
    expect(submitSpy).toBeCalled();
  });

  it('should render a album image', () => {
    const { getByTestId } = renderWithRouterMatch(
      <ITunesDetailContainer dispatchGetTrackDetails={submitSpy} dispatchClearTrackDetails={clearSpy} />,
      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    expect(getByTestId('album-image')).toBeInTheDocument();
  });

  it('should call dispatchClearTracks on unmount', () => {
    const { unmount } = renderWithRouterMatch(
      <ITunesDetailContainer dispatchGetTrackDetails={submitSpy} dispatchClearTrackDetails={clearSpy} />,
      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    unmount();
    expect(clearSpy).toBeCalled();
  });

  it('should call parse the time correctly', () => {
    const { getByTestId } = renderWithRouterMatch(
      <ITunesDetailContainer
        dispatchGetTrackDetails={submitSpy}
        dispatchClearTrackDetails={clearSpy}
        trackDetails={{ trackTimeMillis: 250000 }}
      />,
      {
        route: '/track/123',
        path: '/track/:id'
      }
    );
    const value = getByTestId('parsed-value');
    expect(value.innerHTML).toEqual('4:10');
  });
});
