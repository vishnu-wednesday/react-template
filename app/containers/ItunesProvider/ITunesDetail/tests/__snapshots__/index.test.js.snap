// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ITunesDetails Container Tests should render and match the snapshot 1`] = `
<body>
  <div>
    <div
      class="ITunesDetail__Container-sc-1kdn98s-0 iroCvn"
    >
      <div
        class="ant-row"
        style="margin-left: -8px; margin-right: -8px;"
      >
        <div
          class="ant-col ant-col-xs-24 ant-col-md-14"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <div
            class="ant-card ant-card-bordered ant-card-hoverable ITunesDetail__StyledCard-sc-1kdn98s-1 bbSgik"
          >
            <div
              class="ant-card-body"
            >
              <section
                class="ITunesDetail__ImageAudioSection-sc-1kdn98s-3 bDMRRo"
              >
                <div
                  class="ant-image"
                  data-testid="album-image"
                  style="width: 240px;"
                >
                  <img
                    class="ant-image-img"
                  />
                  <div
                    class="ant-image-mask"
                  >
                    <div
                      class="ant-image-mask-info"
                    >
                      <span
                        aria-label="eye"
                        class="anticon anticon-eye"
                        role="img"
                      >
                        <svg
                          aria-hidden="true"
                          data-icon="eye"
                          fill="currentColor"
                          focusable="false"
                          height="1em"
                          viewBox="64 64 896 896"
                          width="1em"
                        >
                          <path
                            d="M942.2 486.2C847.4 286.5 704.1 186 512 186c-192.2 0-335.4 100.5-430.2 300.3a60.3 60.3 0 000 51.5C176.6 737.5 319.9 838 512 838c192.2 0 335.4-100.5 430.2-300.3 7.7-16.2 7.7-35 0-51.5zM512 766c-161.3 0-279.4-81.8-362.7-254C232.6 339.8 350.7 258 512 258c161.3 0 279.4 81.8 362.7 254C791.5 684.2 673.4 766 512 766zm-4-430c-97.2 0-176 78.8-176 176s78.8 176 176 176 176-78.8 176-176-78.8-176-176-176zm0 288c-61.9 0-112-50.1-112-112s50.1-112 112-112 112 50.1 112 112-50.1 112-112 112z"
                          />
                        </svg>
                      </span>
                      Preview
                    </div>
                  </div>
                </div>
                <div
                  class="sc-bdnxRM hAEZJt"
                >
                  <p
                    class="T__StyledText-gjlic1-0 jbStwN"
                    data-testid="t"
                  >
                    Play it!
                  </p>
                  <audio
                    class="sc-gtsrHT buJEBB"
                    controls=""
                  />
                  <div
                    class="Clickable__StyledClickable-sc-19lp5om-0 grlVIv"
                    data-testid="clickable"
                  >
                    <p
                      class="T__StyledText-gjlic1-0 jbStwN"
                      data-testid="t"
                    >
                      Click here to go to Apple Music store
                    </p>
                  </div>
                </div>
              </section>
              <div
                class="ant-divider ant-divider-horizontal"
                role="separator"
              />
              <section
                class="ITunesDetail__StyledSection-sc-1kdn98s-2"
              >
                <p
                  class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__HeadingT-sc-1kdn98s-4 XzTff"
                  data-testid="t"
                >
                  
                </p>
                <p
                  class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__StyleT-sc-1kdn98s-5 cVklbg"
                  data-testid="t"
                >
                  
                </p>
              </section>
            </div>
          </div>
        </div>
        <div
          class="ant-col ant-col-xs-24 ant-col-md-10"
          style="padding-left: 8px; padding-right: 8px;"
        >
          <div
            class="ant-list ant-list-split ant-list-bordered"
          >
            <div
              class="ant-spin-nested-loading"
            >
              <div
                class="ant-spin-container"
              >
                <li
                  class="ant-list-item"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-avatar"
                    >
                      <span
                        class="ant-avatar ant-avatar-circle"
                        style="margin-top: 5px;"
                      >
                        <span
                          class="ant-avatar-string"
                          style="opacity: 0;"
                        />
                      </span>
                    </div>
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <p
                          class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__ListTitleT-sc-1kdn98s-6 befCwD"
                          data-testid="t"
                        >
                          Primary Genre:
                        </p>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <span
                          class="ant-tag ant-tag-green-inverse"
                        >
                          <p
                            class="ITunesDetail__StyledPara-sc-1kdn98s-7 hqCqyI"
                          />
                        </span>
                      </div>
                    </div>
                  </div>
                </li>
                <li
                  class="ant-list-item"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-avatar"
                    >
                      <span
                        class="ant-avatar ant-avatar-circle"
                        style="margin-top: 5px;"
                      >
                        <span
                          class="ant-avatar-string"
                          style="opacity: 0;"
                        />
                      </span>
                    </div>
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <p
                          class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__ListTitleT-sc-1kdn98s-6 befCwD"
                          data-testid="t"
                        >
                          Collection Name:
                        </p>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <p
                          class="ITunesDetail__StyledPara-sc-1kdn98s-7 hqCqyI"
                        />
                      </div>
                    </div>
                  </div>
                </li>
                <li
                  class="ant-list-item"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-avatar"
                    >
                      <span
                        class="ant-avatar ant-avatar-circle"
                        style="margin-top: 5px;"
                      >
                        <span
                          class="ant-avatar-string"
                          style="opacity: 0;"
                        />
                      </span>
                    </div>
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <p
                          class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__ListTitleT-sc-1kdn98s-6 befCwD"
                          data-testid="t"
                        >
                          Track Duration:
                        </p>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <p
                          class="ITunesDetail__StyledPara-sc-1kdn98s-7 hqCqyI"
                          data-testid="parsed-value"
                        >
                          0:00
                        </p>
                      </div>
                    </div>
                  </div>
                </li>
                <li
                  class="ant-list-item"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-avatar"
                    >
                      <span
                        class="ant-avatar ant-avatar-circle"
                        style="margin-top: 5px;"
                      >
                        <span
                          class="ant-avatar-string"
                          style="opacity: 0;"
                        />
                      </span>
                    </div>
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <p
                          class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__ListTitleT-sc-1kdn98s-6 befCwD"
                          data-testid="t"
                        >
                          Track Explicit:
                        </p>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <p
                          class="ITunesDetail__StyledPara-sc-1kdn98s-7 hqCqyI"
                        />
                      </div>
                    </div>
                  </div>
                </li>
                <li
                  class="ant-list-item"
                >
                  <div
                    class="ant-list-item-meta"
                  >
                    <div
                      class="ant-list-item-meta-avatar"
                    >
                      <span
                        class="ant-avatar ant-avatar-circle"
                        style="margin-top: 5px;"
                      >
                        <span
                          class="ant-avatar-string"
                          style="opacity: 0;"
                        />
                      </span>
                    </div>
                    <div
                      class="ant-list-item-meta-content"
                    >
                      <h4
                        class="ant-list-item-meta-title"
                      >
                        <p
                          class="T__StyledText-gjlic1-0 jbStwN ITunesDetail__ListTitleT-sc-1kdn98s-6 befCwD"
                          data-testid="t"
                        >
                          Year of Release:
                        </p>
                      </h4>
                      <div
                        class="ant-list-item-meta-description"
                      >
                        <p
                          class="ITunesDetail__StyledPara-sc-1kdn98s-7 hqCqyI"
                        >
                          NaN
                        </p>
                      </div>
                    </div>
                  </div>
                </li>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</body>
`;
